pretyping/pretyping.ml

```diff
@@ -870,6 +902,21 @@ struct
           let sigma, hj = pretype empty_tycon env sigma c in
           error_cant_apply_not_functional
             ?loc:(Loc.merge_opt floc argloc) !!env sigma resj [|hj|]
+        in
+        match impls with
+        | None -> f c rest
+        | Some (coe,_) ->
+          let c',rest',some_change = insert_impargs ?loc impls c rest in
+          if not some_change then f c rest else
+          try f c' rest'
+          with e' when CErrors.noncritical e' ->
+          (* Temporary fallback on not using implicit arguments in coercions to funclass *)
+            try
+              let x = f c rest in
+              warn_tolerance_no_coercion_implicit_argument ?loc (!!env,sigma,coe);
+              x
+            with e when CErrors.noncritical e ->
+              raise e'
```

Comment:
Instead of doing this try-with dance, you should return an algebraic type in `f`, catching the exception there. *Using exceptions for control flow is a recipe for a disaster.*

Reply:
@gares, your suggestion to put the quantification in prenex position works, but does not survive composition.
So it seems to me that it is a clear improvement. However, the `gfunctor` file is an infrastructure library and  probably not one as curated as other files in mathcomp. As far as I can tell (but I did not inspect the warnings), this does not impact "client code" of this infrastructure.  So I am not sure that this would bring us a significant gain.

Reply:
Is the risk that `warn` itself raises an error?

I can certainly use a sum type instead, but this is anyway supposed to be temporary and I fail to see how much it really matters here, even though I'm ready to follow recommended common practices which improve our synergy.

Reply:
@ppedrot: I experimented using a sum type. It is actually nicer than what I had thought!